### ISSUE TITLE
Implement MaxMin Diversity based splits

### ISSUE THREAD

**Issue by proteneer** [deepchem/deepchem#889]
Roger Sayle has improved the speed of the maxmin diversity picker in RDKit by like a factor of 100x. Long story short, it allows one to pick a fraction K molecules out of N such that K is maximally diverse from the the N molecules. It also allows for a constrained variant where you have a precursor M that you know _a priori_

It's conceivable that we can use this to generate better splits where give a dataset of N molecules, we:

Maximize the diversity of the test set, i.e. M using the vanilla MaxMin diversity picker. Then,
we set M to be our constrained set, and pick another K to be maximally diverse under M. Everything else, N-K-M, are our training data.

This *may* be better than Murcko-Scaffold decomposition (i.e. no hand written fragmentation rules), and Butina (unequal sized cluster/lump problem).  

Note that this generates test sets that are maximally diverse (which is different maximally dissimilar)

**Comment by rbharath** [comment]
Would definitely love to see a PR for this added in!

**Comment by lilleswing** [comment]
http://www.rdkit.org/docs/GettingStartedInPython.html

What version did they improve the speed in?

**Comment by mwojcikowski** [comment]
2017.09, see https://github.com/rdkit/rdkit/releases/tag/Release_2017_09_1

----
Pozdrawiam,  |  Best regards,
Maciek Wójcikowski
maciek@wojcikowski.pl

2017-10-23 16:58 GMT+02:00 Karl Leswing <notifications@github.com>:

**Comment by rbharath** [comment]
Closing since PR has been merged.
